### DISCUSSION TITLE
Xref a Generic class / method

### DISCUSSION THREAD

**Discussion by RohitM-IN** [statiqdev/Discussions#210] [Q&A]
Hi,

I am new to this wyam & statiq and need help with linking the generic api class 
I have a ref 

```cshtml
[`Result<T>`](xref:api-DbSyncKit.Core.DataContract.Result)
```

how do i get the link of it 

i tried 
```cshtml
xref:api-DbSyncKit.Core.DataContract.Result`1
xref:api-DbSyncKit.Core.DataContract.Result<T>
xref:api-DbSyncKit.Core.DataContract.Result(T)
xref:api-DbSyncKit.Core.DataContract.Result`T
```

this is the class that it refers to 
![image](https://github.com/statiqdev/Discussions/assets/28777412/d4166ee2-3d9b-43f3-a0d4-99be37987cc5)

other are working like `xref:api-DbSyncKit.Core` works perfectly fine

also how do i refrence a property 

DbSyncKit.DB.Attributes.GenerateInsertWithIDAttribute.IncludeIdentityInsert
![image](https://github.com/statiqdev/Discussions/assets/28777412/621423a0-28dc-4658-b536-d7645944528c)

it gives `Couldn't find document with xref "DbSyncKit.DB.Attributes.GenerateInsertWithIDAttribute.IncludeIdentityInsert"`

**Comment by daveaglick** [comment]
This is a great question - it actually made me write some extra tests to verify my understanding about how the code analysis xrefs are generated.

In answer to you first question about generic class xrefs, I'm surprised `xref:api-DbSyncKit.Core.DataContract.Result<T>` isn't working. The unit test I rigged up seems to suggest that's correct:

![image](https://github.com/statiqdev/Discussions/assets/1020407/3e9ffcd5-22c4-4cf2-b65d-0f2989fceabc)

If I had to guess, I'm wondering if the use of angle brackets here is getting in the way since those are relatively loaded as a symbol for web stuff. For example, perhaps they're getting escaped before the xref resolver bits can convert the link. I'm going to try some stuff to get to the bottom of that and will report back. I may end up converting angle brackets to something else like dashes for xref purposes (which I'll need to make sure to document well).

I'm also a little stumped about member xrefs right now. You're correct that those aren't working as intended, but I'm still not sure why. I've got a failing unit test now, so I just need to dig into it and see what's going on. This could get a little tricky though with things like method overloads. If you notice, all members end up with a numerical code that's a hash of their fully qualified name. I might end up using that as a part of the xref, which means you'd need to generate the docs to see what that code is, and then use that. Still thinking about this one.

**Reply by daveaglick** [reply]
Okay, I've run this down and properties and methods do work the same way, though methods also include their arguments (see https://www.statiq.dev/guide/docs/linking-to-apis). In your case...

> it gives Couldn't find document with xref "DbSyncKit.DB.Attributes.GenerateInsertWithIDAttribute.IncludeIdentityInsert"

was on the right track, it looks like you're just missing the "api-" prefix to the xref. All API xrefs have the "api-" prefix to avoid colliding with other xrefs from elsewhere. So in this case, the correct xref is likely "api-DbSyncKit.DB.Attributes.GenerateInsertWithIDAttribute.IncludeIdentityInsert".

![image](https://github.com/statiqdev/Discussions/assets/1020407/a4c26ffc-7ee6-4a51-b48a-fe51a7c62d0d)

Note that I did go ahead and change the xref generation for generic types to convert angle brackets to dashes. This change isn't out yet, but will be rolling out in the Statiq.Docs release 1.0.0-beta.16 later today.

**Reply by RohitM-IN** [reply]
ok that was my mistake but the generic was not solved 
console output
```
 - Couldn't find document with xref "api-DbSyncKit.Core.DataContract.Result%3CT%3E.ResultChangeType"
 - Couldn't find document with xref "api-DbSyncKit.Core.DataContract.Result%3CT%3E.DestinaionDataCount"
 - Couldn't find document with xref "api-DbSyncKit.Core.DataContract.Result%3CT%3E.SourceDataCount"
 - Couldn't find document with xref "api-DbSyncKit.Core.DataContract.Result%3CT%3E.Edited"
 - Couldn't find document with xref "api-DbSyncKit.Core.DataContract.Result%3CT%3E.Deleted"
 - Couldn't find document with xref "api-DbSyncKit.Core.DataContract.Result%3CT%3E.Added"
```

you can check it here [docs page demo](https://dbsynckit.rohit-mahajan.in/guide/usage/synchronization-results )

i have multiple issues here 

1) `xref:api-DbSyncKit.Core.DataContract.Result<T>` which is [here](https://dbsynckit.rohit-mahajan.in/api/dbsynckit.core.datacontract/result_1/)
2) `xref:api-DbSyncKit.Core.DataContract.Result<T>.Added` which is [here](https://dbsynckit.rohit-mahajan.in/api/dbsynckit.core.datacontract/result_1/f6478d76)

so i can tell that the doc is generated but link is just not properly referenced 

Note:
I have updated the `Statiq.Docs` package to `1.0.0-beta.16` 

> Note that I did go ahead and change the xref generation for generic types to convert angle brackets to dashes.

can you provide an example for me

**Reply by RohitM-IN** [reply]
ok nvm i checked commit history and found the change log example 

> For example, the type `RedNs.GreenClass<T>` now gets an xref of `api-RedNs.GreenClass-T-`.

still though i would prefer this to be 

```
 api-RedNs.GreenClass`T
```
 instead as add  as ` will be cleaner no ?

**Reply by daveaglick** [reply]
The full name we get from Roslyn has angle brackets so the easiest thing to do here is replace those angle brackets with something else that doesn't clash with HTML. This matches the full name that the generic type uses on the actual page (except for the replacement of course).

In contrast, the backtick syntax would get confusing, both because the name wouldn't completely match the title of that page being linked to (I.e. the full name with angle brackets), but more importantly because it would be too easy to confuse with the reflection type name convention of using a backtick with the number (not name) of generic type parameters.

**Reply by RohitM-IN** [reply]
can you tell me same for methods 

```csharp
public string GetSqlQueryForSyncData<T>(Result<T> result, int BatchSize = 20) 
    where T : IDataContractComparer
```
namesapce:  `DbSyncKit.Core`

result namespace will be : `DbSyncKit.Core.DataContract`
![image](https://github.com/statiqdev/Discussions/assets/28777412/89eb7870-7dbd-4eda-bd68-bbdc82791bdb)

I tried 
```
 - Couldn't find document with xref "api-DbSyncKit.Core.Synchronization.GetSqlQueryForSyncData-T-"
 - Couldn't find document with xref "api-DbSyncKit.Core.Synchronization.GetSqlQueryForSyncData-T-(DbSyncKit.Core.DataContract.Result-T-,int)"
```

**Reply by daveaglick** [reply]
That's a good question! I would have guessed the second form would work, but it might be a tricky one. I won't be in a position to check until after the New Year (on vacation without a laptop), but I'll take a look once I return.

In the meantime, if you want to take a look, you could add something like `<div>@doc.GetString("xref")</div>` (might not be exactly right). Add that to the layout footer and it'll output the xref for every page in the footer to examine.

**Reply by RohitM-IN** [reply]
guess what it is 

```
api-DbSyncKit.Core.Synchronization.GetSqlQueryForSyncData-T-(DbSyncKit.Core.DataContract.Result-T-, System.Int32)
```

and it outputted as 
![image](https://github.com/statiqdev/Discussions/assets/28777412/e414bd72-79dd-45dd-b229-a2c479db2b1d)

I removed the space between and it changed back to link but now 

```
 - Couldn't find document with xref "api-DbSyncKit.Core.Synchronization.GetSqlQueryForSyncData-T-(DbSyncKit.Core.DataContract.Result-T-,System.Int32)"
```

I guess it will need to remove the space in between or it displays as text XD

---
but man you should add this tip in the docs 

that if you are unsure what the page url looks like just add

```
<div>@Document.GetString("xref")</div>
``` 
to the shared footer and it will give its xref `for that page only` man this is good XD

**Reply by TheBrambleShark** [reply]
Been sort of following this in the background. Using dashes feels strange to me but at the same time, I'm not sure what else I'd use.

HTML uses angle brackets, which is the cause of this issue.
Javadocs uses curly braces, e.g. `<see cref="Task{T}"/>`, but I think this may cause other issues being a reserved character as well.
A single back tick `` ` `` might work, but I worry about it conflicting with Markdown processing.
Square brackets make it look like an accessor rather than a type parameter, so that may work but it'd probably just add confusion.

Suppose we could always use a 😄 emoji `GetSqlQueryForSyncData😄T😄()` /jk

Overall, I think hyphen is probably the best option here.

**Reply by daveaglick** [reply]
> Suppose we could always use a 😄 emoji `GetSqlQueryForSyncData😄T😄()` /jk

Wish I'd thought of that! LGTM.

> Overall, I think hyphen is probably the best option here.

Yeah, I went through a pretty similar thought process as you. Hyphens aren't awesome, but they're less not-awesome than other alternatives I could think of, and in the end it's just for xrefs in specific cases so it's not a thing that's going to be used super frequently.

**Reply by girlpunk** [reply]
You might need opening and closing characters to be different, so _obviously_ the solution is `GetSqlQueryForSyncData👉T👈()`

**Reply by daveaglick** [reply]
> so obviously the solution is `GetSqlQueryForSyncData👉T👈()`

I'm _so_ tempted to hit the "Mark as answer" button on this.
